sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/variant/VariantExpressionEvalUtils.scala

```diff
@@ -41,4 +43,69 @@ object VariantExpressionEvalUtils {
           input.toString, BadRecordException(() => input, cause = e))
     }
   }
+
+  /** Cast a Spark value from `dataType` into the variant type. */
+  def castToVariant(input: Any, dataType: DataType): VariantVal = {
```

Comment:
shall we add overload methods for some primitive types like int, long, double to reduce boxing overhead?

Reply:
I think it is doable but involves quite a lot of boilerplate code. I would prefer to have the optimization in a separate follow-up PR.